src/hotspot/share/utilities/intrusiveList.hpp

```diff
@@ -0,0 +1,1528 @@
+                  typename FromList::const_iterator to) {
+    using Iterator = const_iterator;
+    using Result = iterator;
+    using IOps = Impl::IteratorOperations<Iterator>;
+
+    assert_is_iterator(pos);
+    from_list.assert_is_iterator(from);
+    from_list.assert_is_iterator(to);
+
+    if (from == to) {
+      // Done if empty range.
+      return make_iterator<Result>(pos);
+    } else if (is_same_list(from_list) && (pos == to)) {
+      // Done if already in desired position.
+      return make_iterator_to<Result>(*from);
+    }
+
+    // Adjust sizes if needed.  Only need adjustment if different
+    // lists and at least one of the lists has a constant-time size.
+    if ((_has_size || from_list._has_size) && !is_same_list(from_list)) {
+      difference_type transferring;
+      if (from_list._has_size &&
+          (from == from_list.cbegin()) &&
+          (to == from_list.cend())) {
+        // If from_list has constant-time size() and we're transferring
+        // all of it, we can use that size value to avoid counting the
+        // number of elements being transferred.
+        transferring = *from_list.size_ptr();
+      } else {
+        // Count the number of elements being transferred.
+        transferring = Impl::distance(from, to);
+      }
+      from_list.adjust_size(-transferring);
+      adjust_size(transferring);
+    }
+
+#ifdef ASSERT
+    // Transfer elements to this list, and verify pos not in [from, to).
+    if (is_same_list(from_list)) {
+      for (Iterator i = from; i != to; ++i) {
+        assert(i != pos, "splice range includes destination");
+      }
+    } else {
+      for (Iterator i = from; i != to; ++i) {
+        set_list(*i, &_impl);
+      }
+    }
+#endif // ASSERT
+
+    // to is end of non-empty range, so has a dereferenceable predecessor.
+    Iterator to_pred = --Iterator(to); // Fetch before clobbered
+    // from is dereferenceable since it neither follows nor equals to.
+    const_reference from_value = *from;
+    IOps::iter_attach(IOps::predecessor(from_value), to);
+    IOps::attach(IOps::iter_predecessor(pos), from_value);
+    IOps::attach(*to_pred, pos);
+    return make_iterator_to<Result>(from_value);
+  }
+
+  /**
+   * Transfers all elements of from_list to this list, inserted before
+   * pos.  Returns an iterator referring to the head of the spliced in
+   * range.  Does not invalidate any iterators.
+   *
+   * precondition: pos must be a valid iterator for this list.
+   * precondition: this != &from_list.
+   *
+   * postcondition: iterators referring to elements that were in
+   * from_list are valid iterators for this list rather than
+   * from_list.
+   *
+   * Complexity: constant if either (a) this does not have a
+   * constant-time size() operation, or (b) from_list has a
+   * constant-time size() operation; otherwise O(number of elements
+   * transferred).
+   */
+  template<typename FromList, ENABLE_IF(Impl::IsListType<FromList>::value)>
```

Comment:
Why is this not using `can_splice<FromList>()`?  In some earlier development version doing something like that
produced much worse compiler error messages than doing the more limited test and leaving the compatibility
checking to the inner call to `splice` with range arguments.  But I think that shouldn't be true now, and this
should be changed to use `can_splice`.  But look at the compiler error messages before making that change.

Reply:
Fixed.  Changing it even made the error message better.